packages/flutter/lib/src/widgets/lazy_block.dart

```diff
@@ -0,0 +1,449 @@
+    // viewport. The currentLogicalIndex is the index of the first child that
+    // we don't need.
+
+    if (currentLogicalOffset < endLogicalOffset) {
+      // The last element is visible. We need to update our reckoning of where
+      // the max scroll offset is.
+      _maxScrollOffset = currentLogicalOffset;
+      _startOffsetUpperLimit = double.INFINITY;
+    } else {
+      // The last element is not visible. Ensure that we have enough headroom
+      // so we don't hit the max scroll offset prematurely.
+      _maxScrollOffset = currentLogicalOffset + blockExtent * 2.0;
+      _startOffsetUpperLimit = currentLogicalOffset - blockExtent;
+    }
+
+    // Remove any unneeded children.
+
+    int currentPhysicalIndex = currentLogicalIndex - _firstChildLogicalIndex;
+    final int numberOfRequiredPhysicalChildren = currentPhysicalIndex;
+    while (currentPhysicalIndex < _children.length) {
+      deactivateChild(_children[currentPhysicalIndex]);
+      _children[currentPhysicalIndex] = null;
+      currentPhysicalIndex += 1;
+    }
+    _children.length = numberOfRequiredPhysicalChildren;
+
+    // We now have the correct physical children, each of which has gone through
+    // layout exactly once. We still need to position them correctly. For
+    // simplicity, we position the first physical child at Offset.zero and use
+    // the paintOffset on the render object to adjust the final paint location
+    // of the children.
+
+    Offset currentChildOffset = Offset.zero;
+    child = block.firstChild;
+    while (child != null) {
+      final _LazyBlockParentData childParentData = child.parentData;
+      childParentData.offset = currentChildOffset;
+      currentChildOffset += _getMainAxisOffsetForSize(child.size);
+      child = childParentData.nextSibling;
+    }
+
+    _updatePaintOffset();
+
+    LazyBlockExtentsChangedCallback onExtentsChanged = widget.onExtentsChanged;
+    if (onExtentsChanged != null)
+      onExtentsChanged(_maxScrollOffset - _minScrollOffset, blockExtent, _minScrollOffset);
+  }
+
+  BoxConstraints _getInnerConstraints(BoxConstraints constraints) {
+    switch (widget.direction) {
+      case Axis.horizontal:
+        return new BoxConstraints.tightFor(height: constraints.maxHeight);
+      case Axis.vertical:
+        return new BoxConstraints.tightFor(width: constraints.maxWidth);
+    }
+  }
+
+  double _getMainAxisExtent(Size size) {
+    switch (widget.direction) {
+      case Axis.horizontal:
+        return size.width;
+      case Axis.vertical:
+        return size.height;
+    }
+  }
+
+  Offset _getMainAxisOffsetForSize(Size size) {
+    switch (widget.direction) {
+      case Axis.horizontal:
+        return new Offset(size.width, 0.0);
+      case Axis.vertical:
+        return new Offset(0.0, size.height);
+    }
+  }
+
+  bool _materializeChild(IndexedBuilder builder, int logicalIndex) {
+    int physicalIndex = logicalIndex - _firstChildLogicalIndex;
```

Comment:
I've inlined it.  Originally I had a lot of little functions but ended up inlining most of them.